oscore-cf/src/main/java/org/eclipse/californium/oscore/OSSerializer.java

```diff
@@ -0,0 +1,469 @@
+	 * @param ctx the OSCore context@param code
+	 * @param options the option set
+	 * 
+	 *
+	 * @return the serialized AAD for OSCore
+	 */
+	public static byte[] serializeSendRequestAAD(int version, OSCoreCtx ctx, OptionSet options) {
+		if (version == CoAP.VERSION) {
+			if (ctx != null) {
+				if (options != null) {
+					CBORObject algorithms = CBORObject.NewArray();
+					algorithms.Add(ctx.getAlg().AsCBOR());
+
+					CBORObject aad = CBORObject.NewArray();
+					aad.Add(version);
+					aad.Add(algorithms);
+					aad.Add(ctx.getSenderId());
+					aad.Add(processPartialIV(ctx.getSenderSeq()));
+					return aad.EncodeToBytes();
+				} else {
+					LOGGER.error(Error.OPTIONSET_NULL);
+					throw new IllegalArgumentException(Error.OPTIONSET_NULL);
+				}
+			} else {
+				LOGGER.error(Error.CTX_NULL);
+				throw new IllegalArgumentException(Error.CTX_NULL);
+			}
+		} else {
+			LOGGER.error(Error.WRONG_VERSION_NBR);
+			throw new IllegalArgumentException(Error.WRONG_VERSION_NBR);
+		}
+	}
+
+	/**
+	 * Prepare the additional authenticated data of a received request.
+	 * 
+	 * external_aad = [ ver : uint, alg : int, request_kid : bstr, request_seq :
+	 * bstr, options : bstr]
+	 * 
+	 * @param version the CoAP version number
+	 * @param seq the sent sequence number
+	 * @param ctx the OSCore context@param code
+	 * @param options the option set
+	 * 
+	 *
+	 * @return the serialized AAD for OSCore
+	 */
+	public static byte[] serializeReceiveRequestAAD(int version, int seq, OSCoreCtx ctx, OptionSet options) {
+		if (version == CoAP.VERSION) {
+			if (seq > -1) {
+				if (ctx != null) {
+					if (options != null) {
+						CBORObject algorithms = CBORObject.NewArray();
+						algorithms.Add(ctx.getAlg().AsCBOR());
+
+						CBORObject aad = CBORObject.NewArray();
+						aad.Add(version);
+						aad.Add(algorithms);
+						aad.Add(ctx.getRecipientId());
+						aad.Add(processPartialIV(seq));
+						return aad.EncodeToBytes();
+					} else {
+						LOGGER.error(Error.OPTIONSET_NULL);
+						throw new IllegalArgumentException(Error.OPTIONSET_NULL);
+					}
+				} else {
+					LOGGER.error(Error.CTX_NULL);
+					throw new IllegalArgumentException(Error.CTX_NULL);
+				}
+			} else {
+				LOGGER.error(Error.SEQ_NBR_INVALID);
+				throw new IllegalArgumentException(Error.SEQ_NBR_INVALID);
+			}
+		} else {
+			LOGGER.error(Error.WRONG_VERSION_NBR);
+			throw new IllegalArgumentException(Error.WRONG_VERSION_NBR);
+		}
+	}
+
+	/**
+	 * Prepare the E or I options for encrypting.
+	 * 
+	 * @param writer the datagram writer collecting the necessary bytes
+	 * @param optionSet the original options
+	 * @param isConfidential true will use E options since we are serializing
+	 *            confidential data, false will use I options
+	 * @param request true is request processing, false is response processing
+	 * @param realCode the real code
+	 */
+	private static void writeOptions(DatagramWriter writer, OptionSet optionSet, int realCode) {
```

Comment:
will be obsolete using `DataSerializer.serializeOptionsAndPayload`.

Reply:
In the calling function:

```
public static byte[] serializeConfidentialData(OptionSet options, byte[] payload, int realCode) {
	if (options != null) {
		if (realCode > 0) {
			DatagramWriter writer = new DatagramWriter();
			OptionSet filteredOptions = OptionJuggle.prepareEoptions(options);
			writer.write(realCode, CoAP.MessageFormat.CODE_BITS);
			DataSerializer.serializeOptionsAndPayload(writer, filteredOptions, payload);
			return writer.toByteArray();
		} else {
			LOGGER.severe(OSExceptionHandler.COAP_CODE_INVALID);
			throw new IllegalArgumentException(OSExceptionHandler.COAP_CODE_INVALID);
		}
	} else {
		LOGGER.severe(OSExceptionHandler.OPTIONSET_NULL);
		throw new IllegalArgumentException(OSExceptionHandler.OPTIONSET_NULL);
	}
}
 ```

Reply:
Ha, how stupid of me ;)